src/machine/board_teensy36.go

```diff
@@ -0,0 +1,87 @@
+// +build nxp,mk66f18,teensy36
+
+package machine
+
+import (
+	"device/nxp"
+)
+
+// CPUFrequency returns the frequency of the ARM core clock (180MHz)
+func CPUFrequency() uint32 { return 180000000 }
+
+// ClockFrequency returns the frequency of the external oscillator (16MHz)
+func ClockFrequency() uint32 { return 16000000 }
+
+// LED on the Teensy
+const LED Pin = 13
+
+var pins = []pin{
+	// {bit, control register, gpio register bank}
+	0:  {16, &nxp.PORTB.PCR16, nxp.GPIOB},
```

Comment:
These pins look like Arduino-style pins.
In TinyGo we've solved this differently. The `Pin` values contain the actual pin numbers in whatever form is convenient for the hardware. In this case, it looks like each port has 32 pins so pin 5 maps top port A pin 5, pin 33 to port B pin 2, etc. You can then define pin constants in this board file, such as `machine.D3` for digital pin 3 or however fits with that particular board. That gives readability without sacrificing performance or code density.

Reply:
With the way I've defined the pins, if the pin number is constant, everything optimizes to direct register writes. I've rearranged an instruction, removed a nop, and replaced movt/movw pairs with a single mov, but this is the result of a simple blink program:

```asm
    ; set bit 5 of PORTC PDDR
    mov r4, #0x400FF064
    ldr r0, [r4, #16]
    orr.w r0, r0, #0x20
    str r0, [r4, #16]

    ; configure PORTC PCR5
    mov r0, #0x4004B014
    mov.w r1, #0x144
    str r1, [r0, #0]

    ; set bit 5
    movs r5, #0x20
l:
    str r5, [r4, #0] ; of PORTC PSOR
    bl <time.Sleep>
    str r5, [r4, #4] ; of PORTC PCOR
    bl <time.Sleep>
    b.n l
```

Originally:

```golang

func main() {
	machine.LED.Configure(machine.PinConfig{Mode: machine.PinOutput})

	for {
		machine.LED.Set(true)
		time.Sleep(time.Second)

		machine.LED.Set(false)
		time.Sleep(time.Second)
	}
}
```

Based on exhaustively scanning through the K66 manual's port mux table, I can say that ports have the following pin ranges:

  * A: 0-29
  * B: 0-11, 16-23
  * C: 0-19
  * D: 0-15
  * E: 0-28

Architecturally, each port might be able to have 32 pins, but that's certainly not how it's actually arranged.

Reply:
> With the way I've defined the pins, if the pin number is constant, everything optimizes to direct register writes.

That is likely only true as long as you're only setting one pin: the compiler will inline everything and/or const-propagate the pin. Otherwise the array will likely _not_ be optimized away.

It's fine if there are gaps. You could do this for example:

```go
const (
    PA00 Pin = 0
    PA01 Pin = 1
    ...
    PA29 Pin = 29
    PB00 Pin = 32
    ...
    PB11 Pin = 43
    PB16 Pin = 48
    ...
)
```

You can then use something like this:

```go
func (p Pin) getPortPin() (*nxp.GPIO_Type, uint32) {
    if p < 32 {
        return nxp.GPIOA, uint32(p)
    } else if p < 64 {
        return nxp.GPIOB, uint32(p) - 32
    } // etc
}

// Set the pin to high or low.    
// Warning: only use this on an output pin!    
func (p Pin) Set(high bool) {    
    port, bit := p.getPortPin()    
    if high {    
        port.PSOR.Set(1 << bit)    
    } else {    
        port.PCOR.Set(1 << bit)    
    }    
}    
```

This is also done for other chips, see for example:
https://github.com/tinygo-org/tinygo/blob/master/src/machine/machine_stm32f103xx.go#L35
https://github.com/tinygo-org/tinygo/blob/master/src/machine/machine_stm32f103xx.go#L94

Doing it this way makes sure that such low-level code only needs to be written once. A board using the same chip (or a different chip in the same family) can simply define a few constants without needing to worry about what `nxp.GPIOB` means. There are more advantages: with this separation I was able to simulate some boards on https://play.tinygo.org using the same pin mapping as the boards use on actual hardware.

There is one issue remaining, which is stuff like `&nxp.PORTB.PCR17`. Microchip has a similar issue with their SAM D chips. It's ugly, but I think you can work around it with some `unsafe.Pointer` hacking:

    port := ... // for example nxp.PORTB
    reg := (*volatile.Register32)(unsafe.Pointer(uintptr(unsafe.Pointer(port)) + pin * 4))

Reply:
> 128 (untyped int constant) overflows int8

`Pin` is defined as `int8` in `machine.go`. While the K66 has less than 128 usable pins, if I use your suggested mechanism, I can't use port E without changing the definition of `Pin`.

Reply:
I implemented the change, but it required changing `Pin` to `int16`

Reply:
That may have an effect on many other targets. I think it's best to switch the `machine.Pin` type to `uint8` (and `machine.NoPin` to 0xff) which should also fix issues for a few other chips (like the stm32f7 family that often have a huge number of pins). This should be done in a separate PR, however.

If `uint8` is not enough, the type should instead be chip family specific instead of changing to `int16` for all chips.

Reply:
> This should be done in a separate PR, however.

So this PR should leave out port E for now? Or are you suggesting a separate PR as a prerequisite to this one?

> the type should instead be chip family specific

This seems like the cleanest solution to me. It looks like most low-level types are chip-specific, so it makes a lot of sense to me for the pin type to be chip-specific.

Reply:
>So this PR should leave out port E for now? 

That seems like the simplest option to me.

Reply:
> > This should be done in a separate PR, however.
> 
> So this PR should leave out port E for now? Or are you suggesting a separate PR as a prerequisite to this one?

I intended it as a prerequisite for this PR, but you can also leave out port E for now. There are a few other targets that leave out some pins because of this.

Reply:
I opened #1190, removed port E, and added comment references to the issue.